File: src/ParentsKnownProgenyImputedFamily.cpp
#include <algorithm>
#include <cassert>
#include "../include/ParentsKnownProgenyImputedFamily.h"
#include "../include/VCFGeno.h"
#include "../include/VCFParentsKnownProgenyImputed.h"
#include "../include/VCFProgenyImputed.h"
#include "../include/KnownFamily.h"
#include "../include/ReferenceHaplotype.h"
#include "../include/OptionSmall.h"
#include "../include/common.h"

using namespace std;


//////////////////// ParentsKnownProgenyImputedFamily ////////////////////

// Is the computational cost sufficiently small even when using ref in HMM?
bool ParentsKnownProgenyImputedFamily::is_small(
											const vector<vector<int>>& ref_haps,
											int L, const OptionSmall& op) {
	const size_t	M = ref_haps[0].size();
	const size_t	NH = ref_haps.size();
	if(NH >= 100)
		return false;
	
	// Allow at most one of the four hidden states to change at each transition.
	const double	R = (NH*NH * (8*NH+4)) / op.precision_ratio;
	return R*M < 1e8 && R < 1e5 && L*R*M < 1e9;
}

size_t ParentsKnownProgenyImputedFamily::compute_upper_NH(size_t M, size_t L,
														const OptionSmall& op) {
	size_t	NH = 2;
	while(true) {
		const double	R = (NH*NH * (8*NH+4)) / op.precision_ratio;
		if(!(R*M < 1e8 && R < 1e5 && L*R*M < 1e9))
			break;
		NH += 1;
	}
	return NH - 1;
}

VCFImputable *ParentsKnownProgenyImputedFamily::create_family_vcf(
									const STRVEC& samples,
									const vector<VCFFamilyRecord *>& records,
									int num_families,
									const vector<vector<int>>& ref_haps,
									const VCFSmall *vcf,
									const OptionSmall& op) {
	const size_t	lower_NH = 10;
	const size_t	upper_NH = 20;
	const auto	NH = compute_upper_NH(records.size(), num_families, op);
	if(is_small(ref_haps, static_cast<int>(num_families), op)) {
		return new VCFParentsKnownProgenyImputed(samples, records,
													ref_haps, ref_haps,
													op.map, 0.01, vcf);
	}
	else if(NH >= lower_NH) {
		const int	NH2 = min(upper_NH, NH);
		const auto	gts_mat = VCFFamilyRecord::extract_sample_genotypes(
																	0, records);
		const auto	gts_pat = VCFFamilyRecord::extract_sample_genotypes(
																	1, records);
		const auto	ref_haps_mat =
			ReferenceHaplotype::filter_haplotypes(ref_haps, gts_mat, NH2);
		const auto	ref_haps_pat =
			ReferenceHaplotype::filter_haplotypes(ref_haps, gts_pat, NH2);
		return new VCFParentsKnownProgenyImputed(samples, records,
													ref_haps_mat, ref_haps_pat,
													op.map, 0.01, vcf);
	}
	else {
		// If the amount of calculation is so large, impute only a parent
		return new VCFProgenyImputed(samples, records, ref_haps,
												true, op.map, 0.01, vcf);
	}
}

VCFGenoBase *ParentsKnownProgenyImputedFamily::impute(
									const VCFSmall *orig_vcf,
									const VCFGenoBase *imputed_vcf,
									const vector<const KnownFamily *>& families,
									const vector<STRVEC>& imputed_progenies,
									const vector<vector<int>>& ref_haps,
									const OptionSmall& op) {
	const size_t	L = families.size();
	vector<VCFImputable *>	vcfs;
	vector<VCFFamily *>	vcf_garbage;
	for(size_t i = 0; i < L; ++i) {
		const KnownFamily	*family = families[i];
		const STRVEC	samples { family->get_mat(),
								  family->get_pat(),
								  imputed_progenies[i][0] };
		auto	*vcf = VCFFamily::create_by_two_vcfs(imputed_vcf, 
														orig_vcf, samples);
		auto	*vcf1 = create_family_vcf(samples, vcf->get_family_records(),
													L, ref_haps, orig_vcf, op);
		vcfs.push_back(vcf1);
		vcf->clear_records();
		vcf_garbage.push_back(vcf);
	}
	
	if(vcfs.empty())
		return NULL;
	
	// Small VCFs are heavy to process, so it will be parallelized.
	VCFImputable::impute_VCFs(vcfs, op.num_threads);
	cout << vcfs.size()
			<< " families whose progeny is imputed have been imputed." << endl;
	auto	*new_vcf = VCFImputable::join(vcfs, orig_vcf->get_samples());
	Common::delete_all(vcf_garbage);
	Common::delete_all(vcfs);
	return new_vcf;
}

File: python/ParentsKnownProgenyImputedFamily.py
from __future__ import annotations

# coding: utf-8
# ParentsKnownProgenyImputedFamily.py
# 両親がknownで後代が一つ以上imputeされている家系を補完する

from typing import Optional

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from VCFFamily import *
from VCFProgenyImputed import *
from VCFParentsKnownProgenyImputed import *
from Map import *
from KnownFamily import KnownFamily
from ReferenceHaplotype import filter_haplotypes
from OptionSmall import OptionSmall


#################### ParentsKnownProgenyImputedFamily ####################

def is_small(ref_haps: list[list[int]], L: int, op: OptionSmall) -> bool:
	M = len(ref_haps[0])				# マーカー数
	NH = len(ref_haps)
	# 4つの一つだけ変わってもOK
	R = NH**2 * (8*NH + 4) / op.precision_ratio
	return R * M < 10**8 and R < 10**5 and L * R * M < 10**9

# is_smallが通るギリギリのNHを求める
def compute_upper_NH(family: Family, M: int, L: int, op: OptionSmall) -> int:
	for NH in count(2):
		R = NH**2 * (8*NH + 4) / op.precision_ratio
		if not (R * M < 10**8 and R < 10**5 and L * R * M < 10**9):
			break
	return NH - 1

def impute_family(family: KnownFamily, samples: list[str],
					records: list[VCFFamilyRecord],
					num_families: int, ref_haps: list[list[int]],
					vcf: VCFSmall, op: OptionSmall) -> VCFGenoBase:
	NH = compute_upper_NH(family, len(records), num_families, op)
	if is_small(ref_haps, num_families, op):
		vcf1 = VCFParentsKnownProgenyImputed(samples, records,
												ref_haps, ref_haps,
												op.map, vcf)
		vcf1.impute()
		return vcf1
	elif NH >= 10:
		NH2 = min(20, NH)
		gts_mat = [ r.geno[0] for r in records ]
		gts_pat = [ r.geno[1] for r in records ]
		ref_haps_mat = filter_haplotypes(ref_haps, gts_mat, NH2)
		ref_haps_pat = filter_haplotypes(ref_haps, gts_pat, NH2)
		vcf2 = VCFParentsKnownProgenyImputed(samples, records,
												ref_haps_mat, ref_haps_pat,
												op.map, vcf)
		vcf2.impute()
		return vcf2
	else:
		# 計算量が大きいときは、片親だけ補完する
		new_samples = [samples[0], samples[2]]
		new_records = [ VCFFamilyRecord(r.pos, [r.geno[0], r.geno[2]])
														for r in records ]
		vcf3 = VCFProgenyImputed(new_samples, new_records, ref_haps,
													True, op.map, vcf)
		vcf3.impute()
		# 補完した親だけのVCFにして返す
		vcf_parent = VCFGenoBase.extract_by_samples(vcf3, samples[:1])
		return vcf_parent

def impute(orig_vcf: VCFSmall, merged_vcf: VCFGenoBase,
			families: list[KnownFamily], imputed_progenies: list[list[str]],
			ref_haps: list[list[int]], op: OptionSmall) -> Optional[VCFGeno]:
	vcfs: list[VCFGenoBase] = []
	for i in range(len(families)):
		family = families[i]
		parents = [family.mat, family.pat]
		progeny = imputed_progenies[i][0]
		samples = parents + [progeny]
		vcf = VCFFamily.create_by_two_vcfs(merged_vcf, orig_vcf, samples)
		vcf1 = impute_family(family, samples, vcf.records,
									len(families), ref_haps, orig_vcf, op)
		vcfs.append(vcf1)
	
	if not vcfs:
		return None
	
	print("%d families whose progeny is imputed have been imputed." %
															len(families))
	new_vcf = VCFGeno.join(vcfs, orig_vcf.samples)
	return new_vcf

__all__ = ['impute']
